hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala

```diff
@@ -0,0 +1,195 @@
+    case ir: IR => lower(ir)
+    case tir: TableIR => lower(tir)
+    case mir: MatrixIR => lower(mir)
+  }
+
+  private[this] def lower(ir: IR): IR = {
+    val lowered = lowerChildren(ir).asInstanceOf[IR]
+    assert(lowered.typ == ir.typ)
+    lowered
+  }
+
+  private[this] def lower(tir: TableIR): TableIR = {
+    val lowered = tableRules.applyOrElse(tir, (tir: TableIR) => lowerChildren(tir).asInstanceOf[TableIR])
+    assert(lowered.typ == tir.typ)
+    lowered
+  }
+
+  private[this] def lower(mir: MatrixIR): TableIR = {
+    val lowered = matrixRules.applyOrElse(mir, (mir: MatrixIR) =>
+      CastMatrixToTable(lowerChildren(mir).asInstanceOf[MatrixIR], entriesFieldName, colsFieldName))
+    assert(lowered.typ == loweredType(mir.typ), lowered.typ + "\n" + loweredType(mir.typ))
+    lowered
+  }
+
+  private[this] def lowerChildren(ir: BaseIR): BaseIR = {
+    val loweredChildren = ir.children.map(lower)
+    if ((ir.children, loweredChildren).zipped.forall(_ eq _))
+      ir
+    else {
+      val newChildren = ir.children.zip(loweredChildren).map {
+        case (_: MatrixIR, CastMatrixToTable(childMIR, _, _)) =>
+          childMIR
+        case (mir: MatrixIR, loweredChild: TableIR) =>
+          CastTableToMatrix(
+            loweredChild,
+            entriesFieldName,
+            colsFieldName,
+            mir.typ.colKey)
+        case (_, loweredChild) =>
+          loweredChild
+      }
+      ir.copy(newChildren)
+    }
+  }
+
+  def colVals(tir: TableIR): IR =
+    GetField(Ref("global", tir.typ.globalType), colsFieldName)
+
+  def globals(tir: TableIR): IR =
+    SelectFields(
+      Ref("global", tir.typ.globalType),
+      tir.typ.globalType.fieldNames.diff(FastSeq(colsFieldName)))
+
+  def nCols(tir: TableIR): IR = ArrayLen(colVals(tir))
+
+  def entries(tir: TableIR): IR =
+    GetField(Ref("row", tir.typ.rowType), entriesFieldName)
+
+  def loweredType(
+    typ: MatrixType,
+    entriesFieldName: String = entriesFieldName,
+    colsFieldName: String = colsFieldName
+  ): TableType = TableType(
+    rowType = typ.rvRowType.rename(Map(MatrixType.entriesIdentifier -> entriesFieldName)),
+    key = typ.rowKey,
+    globalType = typ.globalType.appendKey(colsFieldName, TArray(typ.colType)))
+
+  import is.hail.expr.ir.IRBuilder._
+
+  private[this] def matrixRules: PartialFunction[MatrixIR, TableIR] = {
+    case CastTableToMatrix(child, entries, cols, colKey) =>
+      TableRename(lower(child), Map(entries -> entriesFieldName), Map(cols -> colsFieldName))
+
+    case MatrixKeyRowsBy(child, keys, isSorted) =>
+      lower(child).keyBy(keys, isSorted)
+
+    case MatrixFilterRows(child, pred) =>
+      lower(child)
+        .rename(Map(entriesFieldName -> MatrixType.entriesIdentifier))
+        .filter(let (va = 'row,
+                     global = 'global.dropFields(colsField))
+                in pred)
+        .rename(Map(MatrixType.entriesIdentifier -> entriesFieldName))
```

Comment:
I agree completely. I added one simplify rule for `rename` a while ago to let me write something simpler downstream, and I think similar rules should be possible for all IR nodes, which would allow a pushing all renames into the roots of the IR.

Once we have all those rules, they should probably get pulled out into a separate pass, both for efficiency (a top-down pass would be linear, while the "rewrite bottom-up" pattern of `Simplify` would be quadratic), and because that pass would establish such a clear invariant (no rewrite nodes).

The issue of field names in the IR has bothered me for a while, and I think this would be a pretty satisfactory solution. It's on my todo list of IR changes.

I don't think your alternative would work. `entriesField` is a field name, not a variable, and `let` can only bind variables. But for now, I just realized there's an easy fix. `entriesFieldName` is actually equal to `MatrixType.entriesIdentifier`, so the renames are actually trivial. I just generalized the simplify rule that looks for empty renames to look for trivial ones, so these will get simplified away.